controllers/provisionserver_controller.go

```diff
@@ -0,0 +1,310 @@
+
+	if op != controllerutil.OperationResultNone {
+		r.Log.Info(fmt.Sprintf("Deployment %s successfully reconciled - operation: %s", instance.Name, string(op)))
+	}
+
+	// Get the provisioning IP of node hosting the pod belonging to the CR
+	podIP, err := r.getProvisionServerProvisioningIP(instance)
+
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if podIP == "" {
+		r.Log.Info(fmt.Sprintf("Deployment %s pod provisioning IP not yet available, requeuing and waiting", instance.Name))
+		return ctrl.Result{RequeueAfter: time.Second * 30}, err
+	}
+
+	// Update Status
+	instance.Status.LocalImageURL = r.getLocalImageURL(podIP, instance)
+	err = r.Client.Status().Update(context.TODO(), instance)
+
+	if err != nil {
+		r.Log.Error(err, "Failed to update CR status %v")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager - prepare controller for use with operator manager
+func (r *ProvisionServerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// TODO: Myabe use filtering functions here since some resource permissions
+	// are now cluster-scoped?
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&ospdirectorv1beta1.ProvisionServer{}).
+		Owns(&appsv1.Deployment{}).
+		Owns(&corev1.ConfigMap{}).
+		Complete(r)
+}
+
+func (r *ProvisionServerReconciler) httpdConfigMapCreateOrUpdate(instance *ospdirectorv1beta1.ProvisionServer) (controllerutil.OperationResult, error) {
+	cm := provisionserver.HttpdConfigMap(instance)
+
+	op, err := controllerutil.CreateOrUpdate(context.TODO(), r.Client, cm, func() error {
+		err := controllerutil.SetControllerReference(instance, cm, r.Scheme)
+
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	return op, err
+
+}
+
+func (r *ProvisionServerReconciler) deploymentCreateOrUpdate(instance *ospdirectorv1beta1.ProvisionServer) (controllerutil.OperationResult, error) {
+
+	// Get volumes
+	initVolumeMounts := provisionserver.GetInitVolumeMounts(instance.Name)
+	volumeMounts := provisionserver.GetVolumeMounts(instance.Name)
+	volumes := provisionserver.GetVolumes(instance.Name)
+
+	deployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      instance.Name,
+			Namespace: instance.Namespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{"deployment": instance.Name + "-provisionserver-deployment"},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{"deployment": instance.Name + "-provisionserver-deployment"},
+				},
+			},
+		},
+	}
+
+	op, err := controllerutil.CreateOrUpdate(context.TODO(), r.Client, deployment, func() error {
+
+		replicas := int32(1)
+
+		deployment.Spec.Replicas = &replicas
+		deployment.Spec.Template.Spec = corev1.PodSpec{
+			HostNetwork: true,
+			Volumes:     volumes,
+			Containers: []corev1.Container{
+				{
+					Name:         "osp-httpd",
+					Image:        "quay.io/abays/httpd:2.4-alpine",
```

Comment:
should this be a CRD parameter?

Reply:
There's a design discussion to be had around this Apache container and the `downloader` init container image.  The current approach emulates what Metal3 did for their provisioning Apache server.  The `downloader` expects a `.gz` image and creates a compressed `qcow2` image at `/usr/local/apache/htdocs/images/<image name>/compressed-<image name>.qcow2`.  This `/usr/local/apache/htdocs` is then the hard-coded `httpd.conf` default for `DocumentRoot`, which the `httpd:2.4-alpine` container uses to serve the image.  Thus the paths are predictable and can be surfaced in the other operator logic.

We could, of course, make a lot (if not all) of this customizable via the CRD.  It's ugly to be using my quay images right now, but eventually these would come from some official registry in their final form.  I guess I assumed the user would care which is ultimately used, but not be concerned with how its stored or served (aside from choosing the port).  That being said, there may be things I am not considering in an OSP context, as I have simply approached this a variation of the Metal3 OCP implementation for now.